### PR TITLE
 Load files from GCS to Bigquery using BigqueryHook

### PR DESCRIPTION
**Pull request by utkarsharma2**
# Description
## What is the current behavior?
When we run the load_file operator for a table in Bigquery and a file in GCS, the data is first loaded to local from the GCS and then uploaded in Bigquery. Because of this, we require both memory and network bandwidth on the worker node. 

partially closes: #429 

## What is the new behavior?
We have used an optimized path that can directly ingest data from GCS to Bigquery, therefore removing both memory and bandwidth requirements on a worker node. Because of this, the speed and amount of data transfer increase. 

## Does this introduce a breaking change?
No

### Checklist
- [X] Created tests cases
- [X] Add a variable to control the selection of optimized path

### Results
Baseline using `GCSToBigQueryOperator` using [benchmark_gcs_to_bigquery.py](tests/benchmark/dags/benchmark_gcs_to_big_query.py)

|Dataset                                    |Size | Duration(seconds)  |
|-------------------------------------------|-----|--------------------|
|covid_overview/covid_overview_10kb.csv     |10 KB| 5.129522           |
|tate_britain/artist_data_100kb.csv         |100KB| 3.319834           |
|imdb/title_ratings_10mb.csv                |10MB | 5.558414           |
|stackoverflow/stackoverflow_posts_1g.ndjson|1GB  | 85.409014          |
|trimmed/pypi/*                             |5GB  | 48.973093          |

Results post optimization
|Dataset                                    |Size | Duration(seconds)  |
|-------------------------------------------|-----|--------------------|
|covid_overview/covid_overview_10kb.csv     |10 KB| 13.57           |
|tate_britain/artist_data_100kb.csv         |100KB| 16.10          |
|imdb/title_ratings_10mb.csv                |10MB | 19.40         |
|stackoverflow/stackoverflow_posts_1g.ndjson|1GB  | 30.26          |
|trimmed/pypi/*                             |5GB  | 59.90          |

### CODE REVIEW

File: src/astro/databases/google/bigquery.py [hunk 1]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+                target_table=target_table,
+                chunk_size=chunk_size,
+                if_exists=if_exists,
+                **kwargs,
+            )
+            return True
+
+        return False
+
+    def get_project_id_from_conn(self, conn: connection) -> str:
```
Comment (sunank200) [inline src/astro/databases/google/bigquery.py:176]: This can be a static method.

Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:176]: Done.

**Comment by utkarsharma2** [comment]
@sunank200 Added new becnhmaking results in PR.

File: src/astro/databases/google/bigquery.py [hunk 2]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ) -> bool:
+        """
+        Checks if optimised path for transfer between File location to database exists
+        and if it does, it transfers it and returns true else false.
+        """
+        method_name = self.OPTIMIZED_PATHS.get(source_file.location.location_type)
+        if method_name:
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:163]: ok look like you do it here, but I think it would be good to have those as two methods.

as if optimized path is not available, we don't actually do the transfer

File: src/astro/databases/google/bigquery.py [hunk 3]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+        return False
+
+    @staticmethod
+    def get_project_id_from_conn(conn: connection) -> str:
+        """
+        Get project id from conn
+        :param conn: Airflow's connection
+        """
+        if conn.extra and conn.extra.get("project"):
+            return str(conn.extra.get("project"))
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:183]:
```suggestion
            return conn.extra["project"]
```

This might fix it and not need `str()`

Comment (kaxil) [reply src/astro/databases/google/bigquery.py:183]: We already check on L182 that "project" key is present in extra.

Also you might actually want `extra_dejson` not `extra`

File: src/astro/databases/google/bigquery.py [hunk 4]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+        elif conn.host:
+            return str(conn.host)
+        raise ValueError(f"conn_id {conn.conn_id} has no project id.")
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:186]: Host as Project id?

File: src/astro/databases/google/bigquery.py [hunk 5]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+
+    def gs_to_bigquery(
+        self,
+        source_file: File,
+        target_table: Table,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,  # skipcq: PYL-W0613
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:192]: Would we ever use chunk_size in this function?

File: src/astro/databases/google/bigquery.py [hunk 6]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ) -> None:
+        """
+        Transfer data from gcs to bigquery
+        :param source_file: Source file that is used as source of data
+        :param target_table: Table that will be created on the bigquery
+        :param chunk_size: Specify the number of records in each batch to be written at a time
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:200]: Not sure chunk size matters here

File: tests/sql/operators/test_load_file.py [hunk 7]
```diff
@@ -715,3 +717,98 @@ def test_load_file_should_fail_loudly(sample_dag, invalid_path, caplog):
...
+            "gs",
+            "bigquery",
+        ): "astro.databases.google.bigquery.BigqueryDatabase.gs_to_bigquery"
+    }
+
+    source = file_uri.split(":")[0]
+    destination = db.sql_type
+    mock_path = optimised_path_to_method[(source, destination)]
+
+    with mock.patch(mock_path) as gs_to_bigquery:
```
Comment (kaxil) [inline tests/sql/operators/test_load_file.py:760]: If you want to mock the path, why not call the task directly instead of trying to run the DAG?

File: src/astro/databases/base.py [hunk 8]
```diff
@@ -419,3 +419,17 @@ def get_sqlalchemy_template_table_identifier_and_parameter(
             self.get_table_qualified_name(table),
             self.get_table_qualified_name(table),
         )
+
+    def check_optimised_path_and_transfer(
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/databases/base.py:423]: Made the change.

File: src/astro/databases/google/bigquery.py (hunk 2 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:163]: Made the change.

File: src/astro/databases/google/bigquery.py [hunk 9]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+            )
+            return True
+
+        return False
+
+    @staticmethod
+    def get_project_id_from_conn(conn: connection) -> str:
+        """
+        Get project id from conn
+        :param conn: Airflow's connection
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:180]: Done.

File: src/astro/databases/google/bigquery.py (hunk 3 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:183]: Resolved.

File: src/astro/databases/google/bigquery.py (hunk 5 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:192]: I hope not, :) Removing.

File: src/astro/databases/google/bigquery.py [hunk 10]
```diff
@@ -135,3 +146,96 @@ def merge_table(
...
+        self,
+        source_file: File,
+        target_table: Table,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,  # skipcq: PYL-W0613
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ) -> None:
+        """
+        Transfer data from gcs to bigquery
+        :param source_file: Source file that is used as source of data
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:198]: Done.

File: src/astro/databases/google/bigquery.py (hunk 6 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:200]: Removed.

File: src/astro/sql/operators/load_file.py [hunk 11]
```diff
@@ -19,6 +19,7 @@ class LoadFile(BaseOperator):
     :param ndjson_normalize_sep: separator used to normalize nested ndjson.
     :param chunk_size: Specify the number of records in each batch to be written at a time.
     :param if_exists: Overwrite file if exists. Default False.
+    :param optimise_load: variable to control use of optimised path for datatransfer.
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py:22]: Done.

File: src/astro/sql/operators/load_file.py [hunk 12]
```diff
@@ -158,6 +170,7 @@ def load_file(
     task_id: Optional[str] = None,
     if_exists: LoadExistStrategy = "replace",
     ndjson_normalize_sep: str = "_",
+    optimise_load: bool = True,
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py:173]: **use_native_support** sounds good to me

File: tests/sql/operators/test_load_file.py [hunk 13]
```diff
@@ -715,3 +717,98 @@ def test_load_file_should_fail_loudly(sample_dag, invalid_path, caplog):
...
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply tests/sql/operators/test_load_file.py:808]: Updated test case to check the method call and verify the args and kwags passed.

File: tests/sql/operators/test_load_file.py (hunk 7 above)
Comment (utkarsharma2) [reply tests/sql/operators/test_load_file.py]: Done.

File: src/astro/sql/operators/load_file.py [hunk 14]
```diff
@@ -82,13 +85,23 @@ def load_data_to_table(self, input_file: File) -> Table:
...
+                    **self.kwargs,
+                )
+            else:
+                database.load_pandas_dataframe_to_table(
+                    source_dataframe=file.export_to_dataframe(),
+                    target_table=self.output_table,
+                    if_exists=if_exists,
+                    chunk_size=self.chunk_size,
+                )
+                if_exists = "append"
```
Comment (tatiana) [inline src/astro/sql/operators/load_file.py:104]: Why are we setting `if_exists` to append here?

File: tests/sql/operators/test_load_file.py [hunk 15]
```diff
@@ -715,3 +717,129 @@ def test_load_file_should_fail_loudly(sample_dag, invalid_path, caplog):
...
+                "configuration": {
+                    "load": {
+                        "sourceUris": [file_uri],
+                        "destinationTable": {"tableId": test_table.name},
+                    }
+                }
+            },
+            "expected_args": (),
+        }
+    }
```
Comment (tatiana) [inline tests/sql/operators/test_load_file.py:780]: What is the benefit of having many different properties defined in a dictionary, as opposed to having separate variables containing the desired values?
```
source_file_location = "gs"
destination_database = "bigquery"
method_path = "airflow.providers.google.cloud.hooks.bigquery.BigQueryHook.insert_job"
expected_kwargs = {
                "configuration": {
                    "load": {
                        "sourceUris": [file_uri],
                        "destinationTable": {"tableId": test_table.name},
                    }
                }
expected_args = ()
```

File: tests/sql/operators/test_load_file.py [hunk 16]
```diff
@@ -715,3 +717,129 @@ def test_load_file_should_fail_loudly(sample_dag, invalid_path, caplog):
...
+    expected_kwargs = optimised_path_to_method[(source, destination)]["expected_kwargs"]
+    expected_args = optimised_path_to_method[(source, destination)]["expected_args"]
+
+    with mock.patch(mock_path) as method:
+        load_file(
+            input_file=File(file_uri),
+            output_table=test_table,
+        ).operator.execute({})
+        assert method.called
+        assert is_dict_subset(superset=method.call_args.kwargs, subset=expected_kwargs)
```
Comment (tatiana) [inline tests/sql/operators/test_load_file.py:794]: Why are we checking for a subset and not an exact value?

File: src/astro/sql/operators/load_file.py (hunk 14 above)
Comment (tatiana) [inline src/astro/sql/operators/load_file.py:104]: Why are we setting `if_exists` to append here?

File: src/astro/databases/google/bigquery.py [hunk 17]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+        """
+        Transfer data from gcs to bigquery
+
+        :param source_file: Source file that is used as source of data
+        :param target_table: Table that will be created on the bigquery
+        :param if_exists: Overwrite table if exists. Default 'replace'
+        """
+        bq_hook = BigQueryHook(
+            gcp_conn_id=target_table.conn_id,
+        )
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:209]: any reason why we're not using `self.hook`?

File: src/astro/databases/google/bigquery.py [hunk 18]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+
+        if if_exists == "replace":
+            self.drop_table(target_table)
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:212]: We may want to do this regardless if it is an optimised implementation or not. Would there be a higher level function where we could place this?

File: src/astro/databases/google/bigquery.py [hunk 19]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+
+        file_type_to_bq_source_format = {
+            FileType.CSV: "CSV",
+            FileType.NDJSON: "NEWLINE_DELIMITED_JSON",
+            FileType.PARQUET: "PARQUET",
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:217]: What happens if the user had a `FileType.JSON`? Should we raise an exception?

File: src/astro/databases/google/bigquery.py [hunk 20]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+        }
+
+        conn = BaseHook.get_connection(target_table.conn_id)
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:220]: do we need `BaseHook`, or could we be using `self.hook`?

File: src/astro/databases/google/bigquery.py [hunk 21]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+
+        load_job_config = {
+            "sourceUris": [source_file.path],
+            "destinationTable": {
+                "projectId": self.get_project_id_from_conn(conn),
+                "datasetId": target_table.metadata.schema,
+                "tableId": target_table.name,
+            },
+            "createDisposition": "CREATE_IF_NEEDED",
+            "writeDisposition": "WRITE_APPEND",
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:230]:
```
if_exists == "replace" -> WRITE_TRUNCATE
if_exists == "append" -> WRITE_APPEND
```
https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.WriteDisposition.html

File: src/astro/databases/google/bigquery.py [hunk 22]
```diff
@@ -135,3 +146,100 @@ def merge_table(
...
+            "sourceFormat": file_type_to_bq_source_format[source_file.type.name],
+            "autodetect": True,
+        }
+
+        # Since bigquery has other options besides used here, we need to expose them to end user.
```
Comment (tatiana) [inline src/astro/databases/google/bigquery.py:235]: Could you give examples?

File: src/astro/sql/operators/load_file.py [hunk 23]
```diff
@@ -82,13 +85,23 @@ def load_data_to_table(self, input_file: File) -> Table:
...
+                    if_exists=self.if_exists,
+                    **self.kwargs,
+                )
+            else:
+                database.load_pandas_dataframe_to_table(
+                    source_dataframe=file.export_to_dataframe(),
+                    target_table=self.output_table,
+                    if_exists=if_exists,
+                    chunk_size=self.chunk_size,
+                )
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py:103]: @tatiana we can definitely use load_file_to_table. I'll make the change.

Regarding the first point, I think it's lean enough right now. But If you feel strongly about it we can move it to the database.

File: src/astro/sql/operators/load_file.py (hunk 14 above)
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py:104]: @tatiana we are iterating over files and the data needs to be appended in the same table.

File: tests/sql/operators/test_load_file.py (hunk 15 above)
Comment (utkarsharma2) [reply tests/sql/operators/test_load_file.py:780]: @tatiana I think variables are useful when we are reusing them, here this config is only consumed at one location.

File: tests/sql/operators/test_load_file.py (hunk 16 above)
Comment (utkarsharma2) [reply tests/sql/operators/test_load_file.py:794]: @because Matching complete dicts also entail we create the exact same dicts in the test cases that might have a certain logic to them and implementing them might be overkill for test case and as this test case is meant to be generic for all source-destination paths it becomes crucial for it to not have that logic. Instead of that, we can check for certain important args and kwargs and that's why we are using subset logic.

File: src/astro/sql/operators/load_file.py (hunk 14 above)
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py:104]: Probably duplicate comment. Addressed above.

File: src/astro/databases/google/bigquery.py (hunk 19 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:217]: @tatiana @kaxil @sunank200  Nice catch! Yes, we should, and I'll update the code reflecting the same. 

Also, there are two things here
1. I think we should drop the support for the JSON files as they are not used for large data transfer the preferred file type is NDJSON and most DBs actually mean NDJSON when they say JSON.
2. We should also include fileTypes in figuring out optimized paths. Right now we are using only (source, destination) it should be (source, destination, fileType) since there might be a support to ingest some fileType and not others by DBs and we should take that into account.

File: src/astro/databases/google/bigquery.py (hunk 20 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:220]: Done.

File: src/astro/databases/google/bigquery.py (hunk 21 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:230]: @tatiana my bad I can't find the doc but I had the impression that 'WRITE_TRUNCATE' retains the schema of an existing table.  Have made the required changes to code.

File: src/astro/databases/google/bigquery.py (hunk 22 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:236]: Added google link for options.

File: src/astro/databases/google/bigquery.py (hunk 4 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:197]: As per our discussion, removing it since this feature is no longer needed because the google provider >= 6.4.2 take care of this.

File: src/astro/databases/google/bigquery.py (hunk 17 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:213]: Done

File: tests/sql/operators/test_load_file.py (hunk 16 above)
Comment (tatiana) [reply tests/sql/operators/test_load_file.py:791]: @utkarsharma2 seems to make sense, could you give an example of a parameter we don't care about in the test?

File: src/astro/sql/operators/load_file.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tatiana) [reply src/astro/sql/operators/load_file.py]: Hi @utkarsharma2 , yes, please, I'd be very grateful if we could use the original method in the `BaseDatabase` class

File: src/astro/sql/operators/load_file.py (hunk 14 above)
Comment (tatiana) [reply src/astro/sql/operators/load_file.py]: The operator should not have to worry about iterating through files - this should be an internal implementation. from within the Database. Also, it is probably worth adding a comment saying this line is used when we have multiple files.

File: tests/sql/operators/test_load_file.py (hunk 15 above)
Comment (tatiana) [reply tests/sql/operators/test_load_file.py]: I just realised this is a generic method which will be reused across multiple optimised paths, @utkarsharma2 - I understand now :)

File: src/astro/databases/google/bigquery.py (hunk 19 above)
Comment (tatiana) [reply src/astro/databases/google/bigquery.py]: @utkarsharma2:
1. I'm fine dropping support to JSON that is not new-line delimited.
2. I have mixed feelings about this. It would be great if the Astro SDK converts between file types, if necessary, to support the optimised path regardless of the file type.

File: src/astro/sql/operators/load_file.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (kaxil) [reply src/astro/sql/operators/load_file.py]: +1

File: src/astro/databases/google/bigquery.py (hunk 19 above)
Comment (kaxil) [reply src/astro/databases/google/bigquery.py]: 1. I think we discussed on Friday that we will fall-back to the non-optimized path if the filetype isn't supported or raise an error for now and drop the support in future major releases
2. I agree we should include filetype in finding optimized path as doing without it means a) we will either need to fallback to pandas or b) convert the file on the source. So happy for you to include filetype in optimization check. And CSV, NDJSON, PARQUET should be supported by most of the DBs anyways

**Review by kaxil — COMMENTED** [review]
Can you also update the PR title :) -- "Optimise loading files from GCS to Bigquery" -- or something like that

File: src/astro/databases/google/bigquery.py (hunk 19 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py]: Noted. Made the required code changes.

File: src/astro/sql/operators/load_file.py (hunk 14 above)
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py]: Done.

File: src/astro/sql/operators/load_file.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py]: Done.

File: tests/sql/operators/test_load_file.py (hunk 16 above)
Comment (utkarsharma2) [reply tests/sql/operators/test_load_file.py:791]: @tatiana - {'source_file': File(location="gs://dag-authoring/test/8e14a473-3f61-42d6-b903-f9daa22b75c50.csv",type="csv"), 'target_table': Table(conn_id='bigquery', name='_tmp_nomxb3e1i1m5tlxtt3rra55y6tjlap28gibfwgzed587wdo2', metadata=Metadata(schema='tmp_astro', database='astronomer-dag-authoring'), columns=[], temp=True), 'if_exists': 'replace'}

File: .github/ci-test-connections.yaml [hunk 24]
```diff
@@ -31,8 +31,9 @@ connections:
   - conn_id: bigquery
     conn_type: bigquery
     description: null
-    extra: null
-    host: "astronomer-dag-authoring"
+    extra:
+      project: "astronomer-dag-authoring"
```
[the remark this answers is not in the document]
Comment (kaxil) [reply .github/ci-test-connections.yaml:35]: The following is the source code to get the project id
https://github.com/apache/airflow/blob/d9694733cafd9a3d637eb37d5154f0e1e92aadd4/airflow/providers/google/common/hooks/base_google.py#L327-L336

https://github.com/apache/airflow/blob/d9694733cafd9a3d637eb37d5154f0e1e92aadd4/airflow/providers/google/common/hooks/base_google.py#L314-L325

File: src/astro/databases/base.py [hunk 25]
```diff
@@ -220,15 +222,56 @@ def load_file_to_table(
...
+        else:
+            dataframe = source_file.export_to_dataframe()
+            self.load_pandas_dataframe_to_table(
+                dataframe,
+                target_table,
+                if_exists,
+                chunk_size,
+            )
+
+    def load_multiple_file_to_table(
```
Comment (kaxil) [inline src/astro/databases/base.py:246]:
```suggestion
    def load_multiple_files_to_table(
```

File: src/astro/databases/base.py [hunk 26]
```diff
@@ -419,3 +462,25 @@ def get_sqlalchemy_template_table_identifier_and_parameter(
...
+        source_file: File,
+        target_table: Table,
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ):
+        """
+        Checks if optimised path for transfer between File location to database exists
+        and if it does, it transfers it and returns true else false.
+        """
+        pass
```
Comment (kaxil) [inline src/astro/databases/base.py:486]: Would be good to raise an Exception/NotImplemented error instead here

File: src/astro/databases/google/bigquery.py [hunk 27]
```diff
@@ -135,3 +150,96 @@ def merge_table(
...
+        target_table: Table,
+        if_exists: LoadExistStrategy = "replace",
+        **kwargs,
+    ):
+        """
+        Checks if optimised path for transfer between File location to database exists
+        and if it does, it transfers it and returns true else false.
+        """
+        method_name = self.OPTIMIZED_PATHS.get(source_file.location.location_type)
+        if method_name:
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:177]: If a file type isn't supported, it will silently pass here. Should we raise an error, what are your thoughts?

File: src/astro/databases/google/bigquery.py (hunk 17 above)
Comment (kaxil) [reply src/astro/databases/google/bigquery.py:213]: We don't need this do we?

File: src/astro/databases/google/bigquery.py [hunk 28]
```diff
@@ -135,3 +150,96 @@ def merge_table(
...
+        # Since bigquery has other options besides used here, we need to expose them to end user.
+        # https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfigurationLoad
+        load_job_config.update(kwargs)
+
+        job_config = {
+            "jobType": "LOAD",
+            "load": load_job_config,
+            "labels": {"target_table": target_table.name},
+        }
+        bq_hook.insert_job(
```
Comment (kaxil) [inline src/astro/databases/google/bigquery.py:243]: Can we not change this to `self.hook.insert_job` and remove `bq_hook`?

File: src/astro/files/base.py [hunk 29]
```diff
@@ -36,11 +36,11 @@ def __init__(
 
     @property
     def path(self) -> str:
-        return self.location.path
+        return self.location.path  # type: ignore
```
Comment (kaxil) [inline src/astro/files/base.py:39]: We should solve this instead of ignoring it -- can be in a separate PR

Comment (utkarsharma2) [reply src/astro/files/base.py:39]: @kaxil I'll raise a separate PR for this.

File: src/astro/databases/google/bigquery.py (hunk 28 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:243]: Done.

File: src/astro/databases/google/bigquery.py [hunk 30]
```diff
@@ -135,3 +150,96 @@ def merge_table(
...
+        )
+
+        write_disposition_val = {"replace": "WRITE_TRUNCATE", "append": "WRITE_APPEND"}
+
+        conn = self.hook.get_connection(target_table.conn_id)
+
+        load_job_config = {
+            "sourceUris": [source_file.path],
+            "destinationTable": {
+                "projectId": self.get_project_id_from_conn(conn),
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:222]: Done.

File: src/astro/databases/google/bigquery.py (hunk 27 above)
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py:177]: Make sense, added code change.

File: src/astro/databases/base.py (hunk 25 above)
Comment (utkarsharma2) [reply src/astro/databases/base.py:246]: Done.

File: src/astro/databases/google/bigquery.py [hunk 31]
```diff
@@ -135,3 +150,101 @@ def merge_table(
...
+            )
+
+    @staticmethod
+    def get_project_id_from_conn(conn: connection) -> str:
+        """
+        Get project id from conn
+
+        :param conn: Airflow's connection
+        """
+        if conn.extra_dejson and conn.extra_dejson.get("project"):
```
[the remark this answers is not in the document]
Comment (tatiana) [reply src/astro/databases/google/bigquery.py:198]: Where are we using this method?

File: src/astro/sql/operators/load_file.py [hunk 32]
```diff
@@ -76,19 +79,18 @@ def load_data_to_table(self, input_file: File) -> Table:
             ndjson_normalize_sep=self.ndjson_normalize_sep,
             database=database,
         )
-        if_exists = self.if_exists
-        for file in get_files(
+        files = resolve_file_path_pattern(
```
Comment (tatiana) [inline src/astro/sql/operators/load_file.py:82]: @utkarsharma2 , let's use in the operator `database.load_file_to_table`, and let's allow the method   `database.load_file_to_table` to decide if and how to iterate through files if needed. From the Astro Python SDK perspective, `load_file` means wildcard.

As an example, the BigQuery load job supports receiving the wildcard for the files to be matched, so we do not need to iterate through files in this particular path.

Comment (utkarsharma2) [reply src/astro/sql/operators/load_file.py]: Moved the logic into database's base class.

File: src/astro/databases/google/bigquery.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/astro/databases/google/bigquery.py]: We are not using it now. Removed it.

**Review by tatiana — APPROVED** [review]
Thanks, @utkarsharma2 ! I'll use this as a base to extend Ankit's work on Snowflake, so we have consistent interfaces! 🙌 
Feel free to merge once the tests pass!
